### PR TITLE
Added feature to collect links found in HTML

### PR DESCRIPTION
**Pull request by jlandahl**
This change adds an optional feature called "links" which will collect the `href`s of any `a` tag found in the HTML. Links are added to an `HTML.links` vector, and the `url` crate is used to ensure that relative links are processed correctly. The links vector is empty if the feature is not enabled, but this could be changed to not even add the vector if the feature is not enabled.

### CODE REVIEW

**Comment by orottier** [comment]
Hi @jlandahl, thanks for the pull request.
Collecting the links is an interesting feature with not much additional cost. So I will happily include this in the library.

I would like to keep this library usage dead simple so I'm wondering if we should use the feature flag or just collect the links always. The `url` crate is really foundational so I don't think we need to worry about using it.

Another thought that crossed my mind is if `links` should maybe be `Vec<(String, String)>` to also include the anchor text. I will complicate the parser a bit.

What are your thoughts?

**Comment by jlandahl** [comment]
I'm glad you like the addition. I put it in a separate feature to make it as unobtrusive as possible, but if you think it belongs in core functionality I'll just remove the feature parts.

I do like the idea of returning the anchor text as well, but maybe as a separate struct? I'll have a go at that shortly and update the PR.

**Comment by orottier** [comment]
> I'm glad you like the addition. I put it in a separate feature to make it as unobtrusive as possible, but if you think it belongs in core functionality I'll just remove the feature parts.

Yes let's do that 

> I do like the idea of returning the anchor text as well, but maybe as a separate struct? I'll have a go at that shortly and update the PR.

Great!

**Comment by jlandahl** [comment]
Ok, take a look now. I added a `Link` struct, and also noticed that there should probably be a `set_url` on `HTML` so that `url_parsed` stays up to date.

File: src/html.rs [hunk 1]
```diff
@@ -90,6 +97,11 @@ impl HTML {
             .read_from(&mut html.as_bytes())
             .map(|dom| Self::from_dom(dom, url))
     }
+
+    pub fn set_url(&mut self, url: Option<String>) {
```
Comment (orottier) [inline src/html.rs:101]: This should probably not be a public method right? There's no use for updating the URL when parsing is already done

**Comment by orottier** [comment]
Apologies for the delay in response. I think there's 3 things left:
- set_url not pub
- fix the build
- add a simple test (in e.g. the existing html test from_string)

I could pick these up if you wish, just let me know

**Comment by jlandahl** [comment]
I'll get back to this soon, sorry for the delay.

By the way, I'm working on a line-by-line translation of Mozilla's Readability to Rust because the existing Rust ports are not very complete or reliable. You might want to take a look at it when it's ready for the de-HTMLifying part of your crate.

**Comment by orottier** [comment]
Merged via #21 , thanks for the work!

**Pull request closed without merging**